### DISCUSSION TITLE
How to animate a side menu sliding out?

### DISCUSSION THREAD

**Discussion by andrewkdci** [AvaloniaUI/Avalonia#15086] [Q&A] [answered]
I have a project that I'm porting over from WinUI. I have a side menu that slides out when it has content. This slide out is animated. There is also an overlay over the rest of the content area that becomes slightly opaque in order to indicate that this area is no longer the main display area. Here are some pics:

![image](https://github.com/AvaloniaUI/Avalonia/assets/130710084/03f35883-34c0-4b56-9ef5-b65e423d53b8)

![image](https://github.com/AvaloniaUI/Avalonia/assets/130710084/f15e44aa-a559-4e4a-bc19-6b9fb659dee2)

Just to be more specific, here is the Visual State Manager from WinUI:

```
         <VisualStateManager.VisualStateGroups>
            <VisualStateGroup x:Name="ContentGroup">
                <VisualState x:Name="HasContent">
                    <VisualState.StateTriggers>
                        <StateTrigger IsActive="{Binding HasContent}"/>
                    </VisualState.StateTriggers>
                    <VisualState.Setters>
                        <Setter Target="PART_BORDER_Overlay.Visibility" Value="Visible"/>
                        <Setter Target="PART_BORDER_MainContent.Visibility" Value="Visible"/>
                    </VisualState.Setters>
                    <VisualState.Storyboard>
                        <Storyboard>
                            <DoubleAnimation x:Name="PART_ANIMATION_OpenMenu" 
                                             Storyboard.TargetName="PART_TRANSFORM_MainContent"
                                             Storyboard.TargetProperty="X"
                                             From="{x:Bind MenuWidth, Mode=OneWay}"
                                             To="0"
                                             Duration="{StaticResource STYLE_DURATION_MediumSpeedAnimation}">
                                <DoubleAnimation.EasingFunction>
                                    <CircleEase EasingMode="EaseOut"/>
                                </DoubleAnimation.EasingFunction>
                            </DoubleAnimation>
                            <DoubleAnimation Storyboard.TargetName="PART_BORDER_Overlay"
                                             Storyboard.TargetProperty="Opacity"
                                             From="0"
                                             To="0.75"
                                             Duration="{StaticResource STYLE_DURATION_MediumSpeedAnimation}">
                                <DoubleAnimation.EasingFunction>
                                    <ExponentialEase Exponent="7" EasingMode="EaseOut"/>
                                </DoubleAnimation.EasingFunction>
                            </DoubleAnimation>
                        </Storyboard>
                    </VisualState.Storyboard>
                </VisualState>
                <VisualState x:Name="Empty">
                    <VisualState.StateTriggers>
                        <StateTrigger IsActive="{Binding Empty}"/>
                    </VisualState.StateTriggers>
                    <VisualState.Setters>
                        <Setter Target="PART_BORDER_Overlay.Visibility" Value="Collapsed"/>
                        <Setter Target="PART_BORDER_MainContent.Visibility" Value="Collapsed"/>
                    </VisualState.Setters>
                </VisualState>
            </VisualStateGroup>
        </VisualStateManager.VisualStateGroups>
```

As you can see, the animations occur when the menu is opening, but are instantaneous when the menu is closing.

I've been trying to figure this out using Avalonia.Xaml.Behaviors and Avalonia Transitions, but have had little luck. My primary issues are:

1. I want the menu to slide out from the right of the page. In WinUI, I did this by animating the RenderTransform from (-MenuWidth) to 0. I can't figure out how to do this with Avalonia Transitions. How do I START a Transition from a non-zero value?
2. I want the behavior of the Transitions to be asymmetric. When the menu has content, I want the animations to play. When it loses content, I want the Transition to be instantaneous. I am having difficulty figuring out how to do this.
3. Are Transitions the best, most WPF-like thing to be using? If there's some mechanism in Avalonia that is closer to WPF (as opposed to CSS), I would prefer that.

Any help is appreciated. Thanks!

**Comment by thevortexcloud** [comment]
Take a look at how the Material theme does it. They have a few custom controls, and one of which is a slide out menu thing.

https://github.com/AvaloniaCommunity/Material.Avalonia

**Reply by andrewkdci** [reply]
Thanks. I'm looking into Side Sheet, which I think is closest to what I want. But I have yet to make it work.

**Comment by andrewkdci** [comment, marked as answer]
I figured out what I need to do. The primary insight is to set up the element to already have the render transform set up to the initial condition (e.g. `MenuWidth`) before the transition is kicked off. This sounds obvious, but in UWP/WinUI, you can set your animation `From="0"` and `To="MenuWidth"`, so no matter where the render transform is at that time, it will instantaneously jump to 0 and then animate to `MenuWidth`. I can't see how to do this in Avalonia, but if I set up my menu to have `translateX(MenuWidth)` as its default render transform, then setting it to `translateX(0px)` will kick off the transition and animation that I want.

Here is the StyledProperty on my UserControl:

```
public static readonly StyledProperty<double> MenuWidthProperty 
    = AvaloniaProperty.Register<RightSideMenuView, double>(nameof(MinMenuWidth));
```

And here is the Border that serves as the side menu:
```
  <UserControl.Resources>
    <conv:TransformOperationsConverter x:Key="transformOpsConv"/> 
  </UserControl.Resources>
```

```
    <Border x:Name="PART_BORDER_MainContent"
            Grid.Column="1"
            Background="{DynamicResource Brush_LayerBackground}"
            Width="{Binding ElementName=PART_CONTROL_MainContent, Path=Width}"
            Classes.hasContent="{Binding HasContent}"
            IsVisible="{Binding HasContent}">
      <ContentControl x:Name="PART_CONTROL_MainContent"
                      Content="{Binding ActiveMenu}"
                      HorizontalContentAlignment="Stretch"
                      VerticalContentAlignment="Stretch"/>
      
      <Border.Styles>
        <Style Selector="Border">
          <Setter Property="RenderTransform" Value="{Binding RelativeSource={RelativeSource AncestorType=UserControl},
                                                     Path=MenuWidth, 
                                                     Converter={StaticResource transformOpsConv}}"/>
        </Style>
        <Style Selector="Border.hasContent">
          <Setter Property="RenderTransform" Value="translateX(0px)"/>
        </Style>
      </Border.Styles>
      
      <Border.Transitions>
        <Transitions>
          <TransformOperationsTransition Property="RenderTransform" Duration="0:0:0.5"/>
        </Transitions>
      </Border.Transitions>
      
    </Border>
```

Also, I created a value converter so that I can dynamically set the translateX value:

```
    public class TransformOperationsConverter : IValueConverter
    {
        public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
        {
            if(targetType == typeof(ITransform) && value != null && value is double menuWidth)
            {
                return TransformOperations.Parse($"translateX({menuWidth}px)");
            }
            return null;
        }

        public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
        {
            throw new NotImplementedException();
        }
    }
```

The side menu now animates as it slides outwards. 

![SideMenuAnimation](https://github.com/AvaloniaUI/Avalonia/assets/130710084/c53ed19e-4a8f-4dbb-b6e8-e74aa25a5f14)

Additionally, the transition when closing is instantaneous, which is what I want. However, I'm still a little mystified as to why the transition is not applied at this time. It has nothing to do with the IsVisible since it still behaves this way when set IsVisible=True directly. So if anyone has any insight on why this is, please let me know.
